### ISSUE TITLE
stable crd creation fails on 1.17 kind

### ISSUE THREAD

**Issue by clux** [clux/kube-rs#178]
Not sure if this is visible elsewhere, but I'm getting an error on applying the following resource:

```yaml
    apiVersion: apiextensions.k8s.io/v1
    kind: CustomResourceDefinition
    metadata:
      name: foos.clux.dev
    spec:
      group: clux.dev
      names:
        kind: Foo
        plural: foos
        shortNames: []
        singular: foo
      scope: Namespaced
      versions:
        - additionalPrinterColumns: []
          name: v1
          served: true
          storage: true
```

i.e. created on a struct not setting `#[kube(apiextensions = "v1beta1")]`

```rust
[2020-03-08T20:39:50Z INFO  crd_apply] apply error: ErrorResponse { status: "Failure", message: "Incorrect version specified in apply patch. Specified patch version: apiextensions.k8s.io/v1, expected: apiextensions.k8s.io/v1beta1", reason: "BadRequest", code: 400 }
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `400`,
 right: `409`', kube/examples/crd_apply.rs:57:13
```

presumably something is causing kubernetes to expect that to be a v1beta version of the crd even if we're saying v1. Perhaps it's because we're not specifying the [mandatory structural schema](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#specifying-a-structural-schema)?

**Comment by clux** [comment]
can be recreated in #177 on a 0.7.0 [kind cluster](https://github.com/kubernetes-sigs/kind) at cluster version 1.17.

**Comment by clux** [comment]
Running tests against 1.16.8 at the moment so I can deal with this later.

**Comment by teozkr** [comment]
Nevermind, same issue on a clean 1.17.0. Weird..

**Comment by teozkr** [comment]
Ah, the issue is that the URL (from `Api`'s `K`) needs to match the type specified in the YAML manifest. `Api::patch` still takes a raw byte array, so we can't verify that. Changing the `apiexts` `use` to `::v1` (from `::v1beta1`) fixes the issue (or rather, crashes with another error since we don't supply a schema).

Ideally this would be a compile error with a typed `Api::apply`, but to make the transition between the two easier we could make the correct `CustomResourceDefinition` an associated type of the generated CRD.

**Comment by teozkr** [comment]
Then again, maybe it's good to leave this a bit janky, to discourage people from using this pattern outside of these trivial examples.

**Comment by praveenperera** [comment]
In `apiextensions.k8s.io/v1` a `schemas` field is required under `version` could this problem be related to that?

https://github.com/clux/kube-rs/issues/264

https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/

**Comment by teozkr** [comment]
@praveenperera #264 is unrelated to this.

**Comment by clux** [comment]
Are you sure? This might be related if we are in violation of the stable spec. Maybe we just irked by on minikube.

**Comment by teozkr** [comment]
See https://github.com/clux/kube-rs/issues/178#issuecomment-602753634 :P. It's still currently going to fail after that because of #264, but then it will give an error message explaining that reason as well.

**Comment by clux** [comment]
This was fixed in #348
